File: day-1/main.py
#!/usr/bin/env python3

from os import PathLike


def parse_input(filepath: str | PathLike[str]) -> tuple[list[int], list[int]]:
    left_list, right_list = [], []

    with open(filepath, "r") as f:
        for line in f:
            ids = line.split()

            left_list.append(int(ids[0]))
            right_list.append(int(ids[1]))

    return left_list, right_list


def compute_total_distance(left_list: list[int], right_list: list[int]) -> int:
    total_distance = 0
    for left_elm, right_elm in zip(sorted(left_list), sorted(right_list)):
        total_distance += abs(left_elm - right_elm)

    return total_distance


def compute_similarity_score(left_list: list[int], right_list: list[int]) -> int:
    similarity_score = 0
    for id in left_list:
        similarity_score += id * right_list.count(id)

    return similarity_score


def main():
    left_list, right_list = parse_input("input.txt")

    total_distance = compute_total_distance(left_list, right_list)

    print(f"Total distance: {total_distance}")

    similarity_score = compute_similarity_score(left_list, right_list)

    print(f"Similarity score: {similarity_score}")


if __name__ == "__main__":
    main()

File: day-1/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"slices"
	"strconv"
	"strings"
)

func parseInput(filepath string) ([]int, []int, error) {
	file, err := os.Open(filepath)
	if err != nil {
		return nil, nil, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	var leftList, rightList []int

	for scanner.Scan() {
		line := scanner.Text()
		ids := strings.Fields(line)

		leftId, err := strconv.Atoi(ids[0])
		if err != nil {
			return nil, nil, err
		}
		rightId, err := strconv.Atoi(ids[1])
		if err != nil {
			return nil, nil, err
		}

		leftList = append(leftList, leftId)
		rightList = append(rightList, rightId)
	}

	return leftList, rightList, nil
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func computeTotalDistance(leftList, rightList []int) int {
	slices.Sort(leftList)
	slices.Sort(rightList)

	total_distance := 0
	for i := 0; i < min(len(leftList), len(rightList)); i++ {
		total_distance += abs(leftList[i] - rightList[i])
	}

	return total_distance
}

func count(number int, list []int) int {
	count := 0
	for _, elm := range list {
		if elm == number {
			count++
		}
	}
	return count
}

func computeSimilarityScore(leftList, rightList []int) int {
	similarityScore := 0
	for _, id := range leftList {
		similarityScore += id * count(id, rightList)
	}

	return similarityScore
}

func main() {
	leftList, rightList, err := parseInput("input.txt")
	if err != nil {
		panic(err)
	}

	totalDistance := computeTotalDistance(leftList, rightList)

	fmt.Println("Total distance:", totalDistance)

	similarityScore := computeSimilarityScore(leftList, rightList)

	fmt.Println("Similarity score:", similarityScore)
}
